private/bufpkg/bufimage/bufimagemodify/bufimagemodifyv2/bufimagemodifyv2.go

```diff
@@ -15,49 +15,442 @@
 package bufimagemodifyv2
 
 import (
+	"errors"
+
 	"github.com/bufbuild/buf/private/bufpkg/bufimage"
 	"github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagemodify/internal"
 	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/types/descriptorpb"
 )
 
+// Marker markssource SourceCodeInfo_Location indices.
 type Marker interface {
+	// Mark marks the given SourceCodeInfo_Location indices.
 	Mark(bufimage.ImageFile, []int32)
 }
 
+// Sweeper sweeps SourceCodeInfo_Locations.
 type Sweeper interface {
+	// Sweep removes SourceCodeInfo_Locations.
 	Sweep() error
 }
 
+// MarkSweeper marks SourceCodeInfo_Location indices and sweeps source code info
+// to remove marked SourceCodeInfo_Locations.
 type MarkSweeper interface {
 	Marker
 	Sweeper
 }
 
+// NewMarkSweeper returns a MarkSweeper.
 func NewMarkSweeper(image bufimage.Image) MarkSweeper {
-	return nil
+	return newMarkSweeper(image)
+}
+
+// Override describes how to modify a file option, and
+// is passed to ModifyXYZ.
+type Override interface {
+	override()
+}
+
+// NewPrefixOverride returns a new override on prefix.
+func NewPrefixOverride(prefix string) Override {
+	return newPrefixOverride(prefix)
+}
+
+// NewValueOverride returns a new override on value.
+func NewValueOverride[T string | bool | descriptorpb.FileOptions_OptimizeMode](val T) Override {
+	return newValueOverride[T](val)
 }
 
+// ModifyJavaPackage modifies java_package. If override is nil, it uses managed
+// mode's default.
 func ModifyJavaPackage(
 	marker Marker,
 	imageFile bufimage.ImageFile,
-	prefix string,
+	override Override,
 ) error {
+	var javaPackageValue string
+	switch t := override.(type) {
+	case prefixOverride:
+		javaPackageValue = internal.JavaPackageValue(imageFile, t.get())
+	case valueOverride[string]:
```

Comment:
A `value` for a straight `java_package` doesn't make sense - you're saying "for all code generated by this `buf.gen.yaml`, regardless of the package of a given `.proto` file, use exactly this java package value". Same for `go_package`, and others likely.